src/gluonts/torch/model/gnninf/estimator.py

```diff
@@ -0,0 +1,100 @@
+# permissions and limitations under the License.
+
+# If you use this code in your work please cite:
+# Multivariate Time Series Forecasting with Latent Graph Inference (https://arxiv.org/abs/2203.03423)
+
+
+import torch
+from gluonts.dataset.common import Dataset
+from gluonts.model.forecast_generator import DistributionForecastGenerator
+from gluonts.torch.model.predictor import PyTorchPredictor
+from . import trans
+from .module import GNNInfModule
+
+
+class GNNInfEstimator:
+    def __init__(
+        self,
+        freq: str,
+        prediction_length: int,
+        context_length: int,
+        batch_size: int,
+        num_batches_per_epoch: int,
+        epochs: int,
+        lr: float,
+        gnn_name: str,
+        nf: int,
+        gnn_layers: int,
+        device: str='cpu',
+    ) -> None:
+        self.freq = freq
+        self.prediction_length = prediction_length
+        self.context_length = context_length
+        self.batch_size = batch_size
+        self.num_batches_per_epoch = num_batches_per_epoch
+        self.epochs = epochs
+        self.lr = lr
+        self.gnn_name = gnn_name # gnn or bpgnn
+        self.nf = nf
+        self.gnn_layers = gnn_layers
+        self.device = device
+
+    def train_model(self, train_dataset: Dataset, window_loss: int=10) -> GNNInfModule:
+        # Build dataloaders
+        data_loader = trans.wrap_with_dataloader(train_dataset, self.batch_size, self.num_batches_per_epoch,
+                                                 self.prediction_length, self.context_length)
+        n_nodes = len(train_dataset.list_data[0]['target'])
+        net = GNNInfModule(self.freq, self.prediction_length, self.context_length, n_nodes=n_nodes,
+                           gnn_name=self.gnn_name, nf=self.nf, gnn_layers=self.gnn_layers, device=self.device)
+
+        optimizer = torch.optim.Adam(net.parameters(), lr=self.lr)
+        loss_arr = []
+        n_samples = 0
+        net.train()
+        for epoch in range(self.epochs):
+            for idx, batch in enumerate(data_loader):
+                context = batch["past_target"].to(self.device)  # shape (bs, context_length, num_nodes)
+                target = batch["future_target"].to(self.device) # shape (bs, pred_length, num_nodes)
+                optimizer.zero_grad()
+                distr_args, loc, scale = net(context)
+                distr = net.distr_output.distribution(distr_args, loc, scale)
+                loss = -distr.log_prob(target.to(net.device))
+                loss = loss.mean()
+
+                loss.backward()
+                optimizer.step()
+                loss_arr.append(loss.detach().item())
+                n_samples += 1
+                if idx % window_loss == 0:
+                    avg_loss = sum(loss_arr[- window_loss:]) / len(loss_arr[- window_loss:])
+                    print('Epoch %d \t Iteration %d/%d: \t Avg Loss (10 last iterations): %.4f' % (epoch, idx, data_loader.length, avg_loss))
```

Comment:
I think it's better to use a logger for this, see e.g. https://github.com/awslabs/gluon-ts/blob/8f948c65c25ea888da4b6165fc3ec3fa487fb6b6/src/gluonts/mx/trainer/_base.py#L411-L415